ReSwift/CoreTypes/Middleware.swift

```diff
@@ -1,14 +1,99 @@
 //
 //  Middleware.swift
-//  ReSwift
+//  FPExamples
 //
-//  Created by Benji Encz on 12/24/15.
-//  Copyright © 2015 Benjamin Encz. All rights reserved.
+//  Created by Charlotte Tortorella on 25/11/16.
+//  Copyright © 2016 Charlotte Tortorella. All rights reserved.
 //
 
-import Foundation
+/**
+ Middleware is a structure that allows you to modify, filter out and dispatch more
+ actions, before the action being handled reaches the store.
+ */
+public struct Middleware {
+    public typealias DispatchFunction = (Action) -> Void
+    public typealias GetState = () -> StateType
 
-public typealias DispatchFunction = (Action) -> Any
-public typealias GetState = () -> StateType?
-public typealias Middleware =
-    (DispatchFunction?, @escaping GetState) -> (@escaping DispatchFunction) -> DispatchFunction
+    internal let transform: (GetState, DispatchFunction, Action) -> [Action]
+
+    /// Create a blank slate Middleware.
+    public init() {
+        self.transform = { [$2] }
+    }
+
+    /**
+     Initialises the middleware with a transformative function.
+     
+     - parameter transform: The function that will be able to modify passed actions.
+     */
+    internal init(_ transform: @escaping (GetState, DispatchFunction, Action) -> [Action]) {
+        self.transform = transform
+    }
+
+    /**
+     Initialises the middleware by concatenating the transformative functions from
+     the middleware that was passed in.
+     */
+    public init(_ first: Middleware, _ rest: Middleware...) {
+        self = rest.reduce(first) {
+            $0.concat($1)
+        }
+    }
+
+    /// Safe encapsulation of side effects guaranteed 
+    /// not to affect the action being passed through the middleware.
+    public func sideEffect(_ effect: @escaping (GetState, DispatchFunction, Action) -> Void)
+        -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).map {
+                effect(getState, dispatch, $0)
+                return $0
+            }
+        }
+    }
+
+    /// Concatenates the transform function of the passed `Middleware` onto the callee's transform.
+    public func concat(_ other: Middleware) -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).flatMap {
+                other.transform(getState, dispatch, $0)
+            }
+        }
+    }
+
+    /// Concatenates the transform function onto the callee's transform.
+    public func map(_ transform: @escaping (GetState, Action) -> Action) -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).map {
+                transform(getState, $0)
+            }
+        }
+    }
+
+    /// Concatenates the transform function onto the callee's transform.
+    public func increase(_ transform: @escaping (GetState, Action) -> [Action]) -> Middleware {
```

Comment:
Yup I'll fix up the documentation and make it more descriptive.

I chose the name `increase` because it's the antonym of `reduce` and the `increase` function does the inverse of the `reduce` function.

I couldn't think of a better name at the time of naming it and I'm still not sure what else I would call it. Do you have any suggestions?